design/httpproxy-prefix-rewrite.md

```diff
@@ -0,0 +1,262 @@
+This means that multiple routing path prefixes can end up referring to the same HTTPProxy.
+
+To reason about most kinds of request rewriting, the user needs to know the final form of the request.
+In other words, it doesn't make sense to want to rewrite a path prefix unless you have some expectations about the path prefix you will receive.
+This implies that even if the owner of a leaf HTTPProxy doesn't control the root or intermediate documents, they still have fairly concrete expectations around the full URL path to expect.
+
+Although there are a number of ways to conceptualize path prefix rewriting, path regex rewriting requires knowledge of the full request path at the time of writing the regex.
+This is because a regex rewrite can be used to re-order arbitratry portions of the URL path, and this can't be done unless the regex is defined as being applied to the complete path.
+
+## High Level Design
+
+Add a `PathRewrite` field to the HTTPProxy [Route](https://github.com/projectcontour/contour/blob/master/apis/projectcontour/v1/httpproxy.go#L125) object.
+This allows us to specify different kinds of rewrite for URL paths without also needing to specify a specific kind of request rewriting (i.e. we can add host rewrite later).
+
+Locating the `PathRewritePolicy` on the `Route` means that rewriting is controlled by the team that controls the leaf HTTPProxy document.
+This allows a useful initial implementation, but does not prevent adding `PathRewritePolicy` fields higher up the HTTPProxy tree (though allowing that would require detecting and resolving conflicts).
+
+```Go
+type Route struct 
+	...
+
+	// The policy for rewriting the path of the request URL
+	// after the request has been routed to a Service.
+	PathRewrite *PathRewritePolicy `json:"pathRewrite,omitempty"`
+}
+
+```
+
+The `PathRewritePolicy` struct initially only supports path prefix replacement.
+Creating a policy struct allows future addition of other kinds of path rewriting.
+
+
+```Go
+// PathRewritePolicy specifies how a request URL path should be
+// rewritten. This rewriting takes place after a request is routed
+// and has no subsequent effects on the proxy's routing decision.
+// 
+// Exactly one field in this struct may be specified.
+type PathRewritePolicy struct {
+    ReplacePrefix []struct {
+        // +optional
+        Prefix string `json:"prefix,omitempty"`
+
+        Replacement string `json:"replacement"`
+    } `json:"replacePrefix,omitempty"`
+}
+```
+
+`PrefixRewrite` is a slice to allow a service to rewrite multiple path prefixes, depending on the shape of the HTTPProxy tree.
+As discussed above, a HTTPProxy leaf can be reached from multiple HTTPProxy roots via different paths.
+This implies that service owners could reasonably need to rewrite multiple routing paths to a canonical version.
+This, in turn, implies that more than one prefix may need to be rewritten.
```

Comment:
This is true, but I think this overcomplicates the semantics. As I understand them, and forgive me if I'm missing something, the semantics are, if you match this route, the matching prefix -- whatever that is -- is replied with the replacement. Yes, that is hard to use in complicated inclusion semantics, but this feature is rarely used in those circumstances, so we shouldn't need to over fit to that problem.

If there is disagreement on this part we probably need to do some user research. To the best of my understanding this feature is used to "blank out" the path prefix for applications that think they are mounted at "/"

Reply:
> Yes, that is hard to use in complicated inclusion semantics, but this feature is rarely used in those circumstances, so we shouldn't need to over fit to that problem.

My perspective on this is that inclusion is a fundamental part of the `HTTPProxy` API. So we should ensure that other parts of the API work with it in a useful, well-defined way.

> this feature is used to "blank out" the path prefix for applications that think they are mounted at "/"

There is a very limited, IMHO, use case for prefix replacement (see non-goals section). Some applications will be able to use it as you describe, but if they ever emit any links, those links are unlikely to be followable (since nothing rewrites them to match the routing prefix). For simple cases, you can do this; I think a more realistic use case is to rewrite API service urls for deployment or legacy reasons (e.g. the `/api/v1` -> `/v1/api`) example)